### ISSUE TITLE
Don't catch exceptions when debugging. 

### ISSUE THREAD

**Issue by paulthomson** [p-org/PSharp#12]
When debugging, the debugger normally breaks on an uncaught exception. But the PSharp runtime catches all exceptions, reports the stack trace and then exits. It would be nice if this did not happen when debugging.

It looks like there is a way to achieve this:
http://stackoverflow.com/questions/21824597/catch-exception-if-debugger-is-not-attached

You create your try-catch blocks like this:

```
[DebuggerStepThrough]
static public void DoSomeStep()
{
    try
    {                
        DoSomething();
    }
    catch( Exception exception )
    {
        Console.WriteLine( exception.Message );
        if( Debugger.IsAttached == true )
        {
            throw;
        }
    }
}
```

The DebuggerStepThrough attribute ends up causing the debugger to break at the line where the exception was thrown originally (and not at the "throw" in the above method). The downside is that you cannot step through this method (without removing the attribute). However, it looks like you have a few small methods where you catch exceptions, so this seems like it would be fine: Do, ExecuteCurrentStateOnEntry, ExecuteCurrentStateOnExit (for both Machine and Monitor). And I doubt users will be stepping through these methods anyway. 

Some comments suggest this only works if "Enable Just My Code (Managed Only)" is ticked, but I guess it's better than nothing. 

Alternative: Only catch specific exceptions (e.g. TaskCanceledException) and add an [unhandled exception event handler](https://msdn.microsoft.com/en-us/library/system.appdomain.unhandledexception(v=vs.110).aspx) (or something similar).

**Comment by pdeligia** [comment]
Hi Paul, great suggestion -- btw for some reason I did not got a notification for this issue in my email and just saw this -- let me check this out.

**Comment by pdeligia** [comment]
This is now implemented in b32a6cd14283dd49672c920855a26484416a4feb

@paulthomson about what you wrote: "Some comments suggest this only works if "Enable Just My Code (Managed Only)" is ticked, but I guess it's better than nothing." do you know if there is a way to enforce it from the P# side / project? instead of having the user do this? it would be actually cool if the user does not have to get a trace that includes P# internal methods.

**Comment by mattmccutchen-microsoft** [comment]
I was trying to understand how this works.  Some wisdom for the next person:

When Just My Code is enabled, the Visual Studio debugger considers an exception "user-unhandled" and breaks on it if the catch block that will initially catch it is in "non-user code" [(reference)](http://blogs.msdn.com/b/visualstudioalm/archive/2015/01/08/understanding-exceptions-while-debugging-with-visual-studio.aspx), regardless of whether that catch block will rethrow it.  Normally an assembly is considered user code if it is unoptimized and has a PDB [(reference)](https://social.msdn.microsoft.com/Search/en-US?query=just%20my%20code&emptyWatermark=true&ac=2), but the ```DebuggerStepThrough``` attribute ensures that the catch blocks included in b32a6cd are never considered user code, so an exception that would be caught by one of these catch blocks is always considered user-unhandled.

The rethrow is not necessary for the debugger to break on a user-unhandled exception.  It only makes a difference if the user tells the debugger to continue (arguably a strange thing to do), in which case the current machine dies and the rest keep running and maybe hang.  I'd prefer if the runtime would exit as it does when not debugging, which would argue for removing the rethrow.  When Just My Code is disabled (so the debugger doesn't detect a user-unhandled exception), I'm seeing the same behavior with and without the rethrow as if I continued from the user-unhandled exception; it's not as if the rethrow is helping in this case by generating a completely unhandled exception for the debugger to break on either.

**Comment by pdeligia** [comment]
Closing this issue. If further enhancement is needed we can reopen new issue.
